### PR TITLE
Add EventBuffer, BufferedConnection to temporarily buffer Events that…

### PR DESCRIPTION
**Pull request by bretthoerner**
… failed to be sent to the Sentry server.

---
This change is [Reviewable](https://reviewable.io/reviews/getsentry/raven-java/252)

### CODE REVIEW

**Comment by bretthoerner** [comment]
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[raven/src/main/java/com/getsentry/raven/buffer/EventBuffer.java, line 21 at r1](https://reviewable.io:443/reviews/getsentry/raven-java/252#-KRoVKuW3lNc1vFt28pu:-KRoVKuW3lNc1vFt28pv:bi2eokm) ([raw file](https://github.com/getsentry/raven-java/blob/38451b46f40fe0e8996a6709cb9b12a698e61f9a/raven/src/main/java/com/getsentry/raven/buffer/EventBuffer.java#L21)):*
> ```Java
>      * Implementation are expected to flush their buffers to the Sentry server.
>      */
>     void flush();
> ```

From Slack:

I don't call `flush` (attempt to re-send events) anywhere yet. not sure who/where will handle that in the general case.

I feel like whatever it is could be generalized and take an `Iterable<Event>` so that ever `EventBuffer` doesn't have to implement `flush` -- they just need a way to feed their buffered events to something else (each `.next()` call would be expected to *remove* it from the buffer and return it to the "flusher")

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/252)*

**Comment by bretthoerner** [comment]
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/getsentry/raven-java/252#-KRoYTPZhZ25sRFfvwSP:-KRoYTPZhZ25sRFfvwSQ:bk239hk) (no related file):*
I'll add tests, but I want to get your general opinion (plus figure out `flush` -- see below).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/252)*

File: raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+     * event is stored in a file named by its UUID.
+     *
+     * @param event Event to store in add directory
+     */
+    @Override
+    public void add(Event event) {
+        if (getNumStoredEvents() >= maxEvents) {
+            logger.warn("Not adding Event because at least "
+                + Integer.toString(maxEvents) + " events are already stored: " + event.getId());
+            return;
```
Comment (bretthoerner) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: Yeah, I think `BufferStrategy` could be a separate interface so it'd work regardless of the buffer. I think this should probably be a future change though?

Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: Yep. (I agree on this being a future change.)

File: raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            if (bufferDir.isDirectory() && bufferDir.canWrite()) {
+                return new DiskBuffer(bufferDir, maxEvents);
+            } else {
+                logger.error(errMsg);
+            }
+        } catch (Exception e) {
+            logger.error(errMsg, e);
+        }
+
+        return new NoopBuffer();
```
Comment (bretthoerner) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: Yeah, this idea came down from Android. I think you're right about throwing, and we can take a different route in Android (maybe). I'll scrap this stuff.

File: raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns an Iterator of Events that are stored to disk at the point in time this method
+     * is called. Note that files may not deserialize correctly, may be corrupted,
+     * or may be missing by the time we attempt to open them - so some care is taken to
+     * only return valid {@link Event}s.
```
Comment (bretthoerner) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: Yea, that's the "at the point in time this method is called" bit, but I must not have been clear enough. ;)

Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: That makes sense on second reading. 😅  Might help to make this a little more explicit for bad readers like myself.

**Comment by bretthoerner** [comment]
Hm, I see that you've requested changes in GH, is there some way for me to ACK them and say it's ready?

**Comment by tkaemming** [comment]
> Hm, I see that you've requested changes in GH, is there some way for me to ACK them and say it's ready?

Not that I know of, other than doing what you just did. :confused:

File: raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+        String errMsg = "Could not create or write to disk buffer dir: " + bufferDir.getAbsolutePath();
+        try {
+            bufferDir.mkdirs();
+            if (!bufferDir.isDirectory() || !bufferDir.canWrite()) {
+                logger.error(errMsg);
+            }
+        } catch (Exception e) {
+            logger.error(errMsg, e);
+        }
```
Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: Wouldn't we still want this block to throw (and the errors to be logged by the caller?)

File: raven/src/test/java/com/getsentry/raven/buffer/DiskBufferTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+public class DiskBufferTest extends BufferTest {
+
+    private static File BUFFER_DIR = new File("./raven-test-buffer-dir");
+    private int maxEvents = 2;
+
+    private DiskBuffer buffer;
+
+    @BeforeMethod
+    public void setup() {
+        buffer = (DiskBuffer) new DiskBuffer(BUFFER_DIR, maxEvents);
```
Comment (tkaemming) [inline raven/src/test/java/com/getsentry/raven/buffer/DiskBufferTest.java]: Is this cast necessary?

File: raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                return event;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns an Iterator of Events that are stored on disk <b>at the point in time this method
+     * is called</b>. Note that files may not deserialize correctly, may be corrupted,
```
Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: Thanks. 😄

File: raven/src/test/java/com/getsentry/raven/buffer/DiskBufferTest.java (hunk 5 above)
Comment (bretthoerner) [inline raven/src/test/java/com/getsentry/raven/buffer/DiskBufferTest.java]: Oops, leftover from static builder.

File: raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java (hunk 4 above)
Comment (bretthoerner) [inline raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java]: Good catch I meant to change both of these to throw.

**Comment by tkaemming** [comment]
✨  :cake: ✨

**Pull request merged**